### PR TITLE
issue-522,implementation for OpenSearch Egress Rules lifecycle on APIv2

### PR DESCRIPTION
**Pull request by OleksiienkoMykyta**
…

### CODE REVIEW

File: pkg/instaclustr/config.go [hunk 1]
```diff
@@ -52,6 +52,8 @@ const (
 	AWSEncryptionKeyEndpoint               = "/cluster-management/v2/resources/providers/aws/encryption-keys/v2/"
 	ListAppsVersionsEndpoint               = "%s/cluster-management/v2/data-sources/applications/%s/versions/v2/"
 	ClusterSettingsEndpoint                = "%s/cluster-management/v2/operations/clusters/v2/%s/change-settings/v2"
+	OpenSearchEgressRulesEndpoint          = "/cluster-management/v2/resources/applications/opensearch/egress-rules/v2"
+	OpenSearchEgressRuleDeleteEndpoint = "/cluster-management/v2/resources/applications/opensearch/egress-rules/v2/"
```
Comment (worryg0d) [inline pkg/instaclustr/config.go:56]: Both constants are identical. Please delete the second one.

File: controllers/clusterresources/suite_test.go [hunk 2]
```diff
@@ -17,35 +17,29 @@ limitations under the License.
...
 )
 
-var (
-	cfg         *rest.Config
-	k8sClient   client.Client
-	testEnv     *envtest.Environment
-	ctx         context.Context
-	cancel      context.CancelFunc
-	MockInstAPI = mock.NewInstAPI()
-)
```
Comment (worryg0d) [inline controllers/clusterresources/suite_test.go:48]: Please revert these changes of suit_test.

**Review by worryg0d — CHANGES_REQUESTED** [review]

File: pkg/instaclustr/config.go (hunk 1 above)
Comment (OleksiienkoMykyta) [reply pkg/instaclustr/config.go:56]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+// OpenSearchEgressRulesList contains a list of OpenSearchEgressRules
+type OpenSearchEgressRulesList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []OpenSearchEgressRules `json:"items"`
+}
+
+func (er *OpenSearchEgressRules) GetJobID(jobName string) string {
+	return client.ObjectKeyFromObject(er).String() + "/" + jobName
+}
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_types.go:60]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OpenSearchEgressRulesSpec struct {
+	ClusterID           string `json:"clusterId"`
+	Name                string `json:"name"`
+	OpenSearchBindingID string `json:"openSearchBindingId"`
+	Source              string `json:"source"`
+	Type                string `json:"type"`
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_types.go:29]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+//+kubebuilder:webhook:path=/validate-clusterresources-instaclustr-com-v1beta1-opensearchegressrules,mutating=false,failurePolicy=fail,sideEffects=None,groups=clusterresources.instaclustr.com,resources=opensearchegressrules,verbs=create;update,versions=v1beta1,name=vopensearchegressrules.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &OpenSearchEgressRules{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *OpenSearchEgressRules) ValidateCreate() error {
+	opensearchegressruleslog.Info("validate create", "name", r.Name)
+
+	if r.Spec.ClusterID == "" || r.Spec.OpenSearchBindingID == "" || r.Spec.Source == "" {
+		return fmt.Errorf("spec.ClusterID, spec.OpenSearchBindingId, spec.Source must be filled")
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: fixed

File: controllers/clusterresources/suite_test.go (hunk 2 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/suite_test.go:48]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OpenSearchEgressRulesSpec struct {
+	ClusterID           string `json:"clusterId"`
+	Name                string `json:"name"`
```
Comment (worryg0d) [inline apis/clusterresources/v1beta1/opensearchegressrules_types.go:26]: As Instaclustr API [doc](https://instaclustr.redoc.ly/Current/tag/OpenSearch-Egress-Rules-V2#paths/~1cluster-management~1v2~1resources~1applications~1opensearch~1egress-rules~1v2/post) says the field `name` should be returned by the API if you add a new egress rule, but this is not part of a creation request. You should move this field from the resource spec to its status

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	OpenSearchBindingID string `json:"openSearchBindingId"`
+	Source              string `json:"source"`
+	Type                string `json:"type,omitempty"`
+}
+
+type OpenSearchEgressRulesStatus struct {
+	ID     string `json:"id,omitempty"`
+	Status string `json:"status,omitempty"`
```
Comment (worryg0d) [inline apis/clusterresources/v1beta1/opensearchegressrules_types.go:34]: This field is never used. Please delete it

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go (hunk 6 above)
Comment (worryg0d) [reply apis/clusterresources/v1beta1/opensearchegressrules_types.go:26]: I've looked at the request sample in the doc, and the `name` field` is passed there. I think you should test what if you don't pass it. If any error is returned then change nothing

**Review by worryg0d — CHANGES_REQUESTED** [review]

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go (hunk 6 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_types.go:26]: It`s described in spec file, and in request body, in general this field does not affect nothing, but I think we should implement code according to spec

File: apis/clusterresources/v1beta1/opensearchegressrules_types.go (hunk 7 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_types.go]: fixed

**Review by worryg0d — APPROVED** [review]

**Review by tengu-alt — APPROVED** [review]

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+var _ webhook.Validator = &OpenSearchEgressRules{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *OpenSearchEgressRules) ValidateCreate() error {
+	opensearchegressruleslog.Info("validate create", "name", r.Name)
+
+	if r.Spec.ClusterID == "" || r.Spec.OpenSearchBindingID == "" || r.Spec.Source == "" {
+		return fmt.Errorf("clusterId, openSearchBindingId, source must be filled")
+	}
```
Comment (testisnullus) [inline apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: Let's remove this code. All these fields are already marked as required in the resource spec

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	clusterresourcesv1beta1 "github.com/instaclustr/operator/apis/clusterresources/v1beta1"
```
Comment (testisnullus) [inline controllers/clusterresources/opensearchegressrules_controller.go:40]: Sort imports please

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+	patch := rule.NewPatch()
+	err = json.Unmarshal(b, &rule.Status)
+	if err != nil {
+		l.Error(err, "failed to parse OpenSearch Egress Rule resource response from Instaclustr")
+		r.EventRecorder.Eventf(rule, models.Warning, models.ConvertionFailed,
+			"Failed to parse OpenSearch Egress Rule response from Instaclustr. Reason: %v", err,
+		)
+
+		return err
```
Comment (testisnullus) [inline controllers/clusterresources/opensearchegressrules_controller.go:125]: Let's move unmarshal logic inside the client method

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	// This hack was added because instaclustr api returns null value egress rule id, but it should be equal to pattern {clusterId}~{source}~{bindingId}
+	// This code could be removed after instaclustr fix bugs, to make code cleaner
```
Comment (testisnullus) [inline controllers/clusterresources/opensearchegressrules_controller.go:129]: // This hack was added because the Instaclustr API returns an egress rule id with a null value, but it should be equal to the pattern {clusterId}~{source}~{bindingId}
// This code could be removed after the Instaclustr API team fixes this bug

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 8 above)
Comment (DoodgeMatvey) [inline apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: You also need to add enum validation for `source`: `"ALERTING"`, `"NOTIFICATIONS"`
And for `type`: `"SLACK"`, `"WEBHOOK"`, `"CUSTOM_WEBHOOK"`, `"CHIME"`

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+// log is for logging in this package.
+var opensearchegressruleslog = logf.Log.WithName("opensearchegressrules-resource")
+
+var openSearchBindingIDPattern, _ = regexp.Compile(`[\w-]+`)
```
Comment (DoodgeMatvey) [inline apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:34]: Please add this regular expression to constants

**Review by DoodgeMatvey — DISMISSED** [review]

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 8 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: I think it`s unnecessary, all such validations implemented on Instaclustr API, and in case they add new types, logic wouldn't work. About "ALERTING" it`s already deprecated, I`d rather avoid such validations if it's possible

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 12 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:34]: It`s only one usage, and it`s very specific regular expression, I don`t think that it is necessary

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: As Danylo sad, this validation is redundant, so I just removed it.
"Let's remove this code. All these fields are already marked as required in the resource spec"

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	if r.Status.ID == "" {
+		return r.ValidateCreate()
+	}
+
+	if r.Spec != oldRules.Spec {
+		return models.ErrImmutableSpec
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:76]: But there are no API calls to update entity, only creation when resource has wrong configuration, to avoid additional operations, like deletion and creation new one, we just can edit resorce and apply it.

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 8 above)
Comment (DoodgeMatvey) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:52]: Yea, but in such way it will be created in k8s but wasn't created in Instaclustr.

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+		if k8serrors.IsNotFound(err) {
+			return models.ExitReconcile, nil
+		}
+
+		l.Error(err, "Unable to fetch OpenSearch Egress Rules resource")
+
+		return models.ReconcileRequeue, err
+	}
+
+	// Handle resource deletion
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go:83]: It`s not TODO, it`s just decription to operations, changed it on "It`s handling resource deletion", to add more context

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 8 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 9 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 10 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:72]: ok

[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:72]: We need to decline the update of this resource. Without validation, if the customer tries to update this entity inside K8s and this action is passed successfully, it can be really confusing if the entity will not be updated on the Instaclustr API, but updated inside K8s

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 12 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+		"OpenSearch Egress Rule has been created",
+	)
+
+	return nil
+}
+
+func (r *OpenSearchEgressRulesReconciler) handleDelete(ctx context.Context, logger logr.Logger, rule *clusterresourcesv1beta1.OpenSearchEgressRules) error {
+	err := r.API.DeleteOpenSearchEgressRule(rule.Status.ID)
+	if err != nil {
+		logger.Error(err, "failed to delete OpenSearch Egress Rule on Instaclustr")
```
Comment (ribaraka) [inline controllers/clusterresources/opensearchegressrules_controller.go:153]: I think it's better to handle the NotFound error

File: pkg/instaclustr/client.go [hunk 16]
```diff
@@ -2291,3 +2291,57 @@ func (c *Client) DeleteAWSEndpointServicePrincipal(principalID string) error {
...
+		return err
+	}
+
+	if resp.StatusCode != http.StatusAccepted {
+		return fmt.Errorf("status code: %d, message: %s", resp.StatusCode, b)
+	}
+	err = json.Unmarshal(b, &rule.Status)
+	if err != nil {
+		return err
+	}
```
Comment (ribaraka) [inline pkg/instaclustr/client.go:2320]: It would be clearer if you just returned the bytes and only then converted them to `rule.Status`. This shows that the status is indeed changing and you are performing the `patch` operation.

File: pkg/instaclustr/interfaces.go [hunk 17]
```diff
@@ -98,6 +98,8 @@ type API interface {
 	ListAppVersions(app string) ([]*models.AppVersions, error)
 	GetDefaultCredentialsV1(clusterID string) (string, string, error)
 	UpdateClusterSettings(clusterID string, settings *models.ClusterSettings) error
+	CreateOpenSearchEgressRules(rule *clusterresourcesv1beta1.OpenSearchEgressRules) error
+	DeleteOpenSearchEgressRule(egressRuleId string) error
```
Comment (ribaraka) [inline pkg/instaclustr/interfaces.go:102]: egressRuleId -> ID

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: pkg/models/errors.go [hunk 18]
```diff
@@ -62,4 +62,5 @@ var (
 	ErrPrivateLinkOnlyWithPrivateNetworkCluster = errors.New("private link is available only for private network clusters")
 	ErrPrivateLinkSupportedOnlyForSingleDC      = errors.New("private link is only supported for a single data centre")
 	ErrPrivateLinkSupportedOnlyForAWS           = errors.New("private link is supported only for an AWS cloud provider")
+	ErrImmutableSpec                            = errors.New("spec is immutable")
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply pkg/models/errors.go]: fixed

File: pkg/instaclustr/client.go (hunk 16 above)
Comment (OleksiienkoMykyta) [reply pkg/instaclustr/client.go:2352]: After discussion with the team, we agreed that we should convert data in client

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 15 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+	opensearchegressruleslog.Info("validate create", "name", r.Name)
+	equalsToReg, _ := regexp.MatchString(models.OpenSearchBindingIDPattern, r.Spec.OpenSearchBindingID)
+
+	if !slices.Contains(source, r.Spec.Source) || !slices.Contains(theType, r.Spec.Type) {
+		return fmt.Errorf("the source should be equeal to one of options: %q , got: %q. the type should be equeal to one of options: %q , got: %q", source, theType, r.Spec.Source, r.Spec.Type)
+	}
+
+	if !equalsToReg {
+		return fmt.Errorf("mismatching openSearchBindingId to pattern: %s", models.OpenSearchBindingIDPattern)
+	}
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+// log is for logging in this package.
+
+var opensearchegressruleslog = logf.Log.WithName("opensearchegressrules-resource")
+
+var theType = []string{"SLACK", "WEBHOOK", "CUSTOM_WEBHOOK", "CHIME"}
+var source = []string{"NOTIFICATIONS"}
```
[the remark this answers is not in the document]
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+			)
+
+			return err
+		}
+	}
+
+	patch := rule.NewPatch()
+
+	// This hack was added because the Instaclustr API returns an egress rule id with a null value, but it should be equal to the pattern {clusterId}~{source}~{bindingId}
+	// This code could be removed after the Instaclustr API team fixes this bug
```
Comment (testisnullus) [inline controllers/clusterresources/opensearchegressrules_controller.go:133]: Remove this comment or move it in the right place

File: pkg/instaclustr/client.go [hunk 22]
```diff
@@ -2320,3 +2320,92 @@ func (c *Client) GetAWSVPCPeering(peerID string) (*models.AWSVPCPeering, error)
...
+
+	return &rule, nil
+}
+
+func (c *Client) DeleteOpenSearchEgressRule(id string) error {
+	// https://api.dev.instaclustr.com/cluster-management/v2/resources/applications/opensearch/egress-rules/v2/{egressRuleId}
+	fmt.Println(id)
+	url := c.serverHostname + OpenSearchEgressRulesEndpoint + "/" + id
+	fmt.Println("GOT:", url)
+	fmt.Println("expected: https://api.dev.instaclustr.com/cluster-management/v2/resources/applications/opensearch/egress-rules/v2/{egressRuleId}")
```
Comment (testisnullus) [inline pkg/instaclustr/client.go:2389]: Remove unnecessary code please

**Review by testisnullus — APPROVED** [review]

File: pkg/instaclustr/client.go (hunk 22 above)
Comment (OleksiienkoMykyta) [reply pkg/instaclustr/client.go]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 21 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+	if err != nil {
+		return fmt.Errorf("can`t match openSearchBindingId to pattern: %s, error: %w", models.OpenSearchBindingIDPattern, err)
+	}
+	if !equalsToReg {
+		return fmt.Errorf("mismatching openSearchBindingId to pattern: %s", models.OpenSearchBindingIDPattern)
+	}
+
+	if !slices.Contains(sourcePlugins, r.Spec.Source) || !slices.Contains(destinationTypes, r.Spec.Type) {
+		return fmt.Errorf("the source should be equeal to one of options: %q , got: %q. the type should be equeal to one of options: %q , got: %q", sourcePlugins, destinationTypes, r.Spec.Source, r.Spec.Type)
+	}
```
Comment (worryg0d) [inline apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:62]: In my opinion, it will be better if you split it into 2 different conditions because the returning error looks unclear to me.
```suggestion
	if !slices.Contains(sourcePlugins, r.Spec.Source) {
		return fmt.Errorf("the source should be equal to one of the options: %q , got: %q", sourcePlugins, r.Spec.Source)
	}
	
	if  !slices.Contains(destinationTypes, r.Spec.Type) {
		return fmt.Errorf("the type should be equal to one of the options: %q , got: %q", destinationTypes, r.Spec.Type
	}
```

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+*/
+
+package clusterresources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
```
Comment (worryg0d) [inline controllers/clusterresources/opensearchegressrules_controller.go:24]: I don't want to say it but anyway someone will say So, please sort imports

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+				"Failed to create OpenSearch Egress Rule on Instaclustr. Reason: %v", err,
+			)
+
+			return err
+		}
+	}
+
+	patch := rule.NewPatch()
+
+	err = r.Status().Patch(ctx, rule, patch)
```
Comment (worryg0d) [inline controllers/clusterresources/opensearchegressrules_controller.go:132]: It looks a bit weird. As I understand here you patch an ID of the resource. I suggest you create another variable for self-composed ID and provide it anywhere you want. Then patch `rule.Status.ID` with the value of the created variable.

File: pkg/instaclustr/client.go [hunk 26]
```diff
@@ -2320,3 +2320,89 @@ func (c *Client) GetAWSVPCPeering(peerID string) (*models.AWSVPCPeering, error)
...
+	defer resp.Body.Close()
+	b, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusAccepted {
+		return fmt.Errorf("status code: %d, message: %s", resp.StatusCode, b)
+	}
+	err = json.Unmarshal(b, &rule.Status)
```
Comment (worryg0d) [inline pkg/instaclustr/client.go:2346]:
```suggestion
	}
	
	err = json.Unmarshal(b, &rule.Status)
```

File: pkg/instaclustr/client.go [hunk 27]
```diff
@@ -2320,3 +2320,89 @@ func (c *Client) GetAWSVPCPeering(peerID string) (*models.AWSVPCPeering, error)
...
+	err = json.Unmarshal(b, &rule)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rule, nil
+}
+
+func (c *Client) DeleteOpenSearchEgressRule(id string) error {
+	// https://api.dev.instaclustr.com/cluster-management/v2/resources/applications/opensearch/egress-rules/v2/{egressRuleId}
```
Comment (worryg0d) [inline pkg/instaclustr/client.go:2385]: Please delete this

File: pkg/instaclustr/client.go [hunk 28]
```diff
@@ -2320,3 +2320,89 @@ func (c *Client) GetAWSVPCPeering(peerID string) (*models.AWSVPCPeering, error)
...
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, NotFound
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("status code: %d, message: %s", resp.StatusCode, b)
+	}
+	err = json.Unmarshal(b, &rule)
```
Comment (worryg0d) [inline pkg/instaclustr/client.go:2376]: A bit of code cleanup. Also do it in other similar cases if they exists
```suggestion
	}
	
	err = json.Unmarshal(b, &rule)
```

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 23 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:62]: fixed

Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:62]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 25 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go:132]: fixed

File: pkg/instaclustr/client.go (hunk 27 above)
Comment (OleksiienkoMykyta) [reply pkg/instaclustr/client.go:2385]: fixed

File: pkg/instaclustr/client.go (hunk 26 above)
Comment (OleksiienkoMykyta) [reply pkg/instaclustr/client.go:2346]: fixed

File: apis/clusterresources/v1beta1/opensearchegressrules_webhook.go (hunk 23 above)
Comment (OleksiienkoMykyta) [reply apis/clusterresources/v1beta1/opensearchegressrules_webhook.go:62]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 24 above)
Comment (OleksiienkoMykyta) [reply controllers/clusterresources/opensearchegressrules_controller.go:24]: fixed

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+	"github.com/go-logr/logr"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
```
Comment (ribaraka) [inline controllers/clusterresources/opensearchegressrules_controller.go:38]: instaclustr/operator imports go third.

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 30]
```diff
@@ -0,0 +1,210 @@
...
+		Name:      req.Name,
+	}, rule)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return models.ExitReconcile, nil
+		}
+
+		l.Error(err, "Unable to fetch OpenSearch Egress Rules resource")
+
+		return models.ReconcileRequeue, err
```
Comment (ribaraka) [inline controllers/clusterresources/opensearchegressrules_controller.go:77]: please apply the custom rate limiter and remove `models.ReconcileRequeue`, leaving only errors.

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusterresources/opensearchegressrules_controller.go (hunk 30 above)
Comment (ribaraka) [reply controllers/clusterresources/opensearchegressrules_controller.go:77]: please apply the custom rate limiter and remove models.ReconcileRequeue

File: controllers/clusterresources/opensearchegressrules_controller.go [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	// It`s handling resource deletion
+	if rule.DeletionTimestamp != nil {
+		err = r.handleDelete(ctx, l, rule)
+		if err != nil {
+			return models.ReconcileRequeue, err
```
Comment (ribaraka) [inline controllers/clusterresources/opensearchegressrules_controller.go:84]: please apply the custom rate limiter and remove `models.ReconcileRequeue`

**Review by ribaraka — APPROVED** [review]
